common/src/main/java/io/netty/util/concurrent/AutoScalingEventExecutorChooserFactory.java

```diff
@@ -0,0 +1,362 @@
+        }
+        if (windowUnit == null) {
+            throw new NullPointerException("windowUnit");
+        }
+        if (scaleDownThreshold < 0.0 || scaleDownThreshold > 1.0) {
+            throw new IllegalArgumentException(
+                    "scaleDownThreshold must be between 0.0 and 1.0: " + scaleDownThreshold);
+        }
+        if (scaleUpThreshold < 0.0 || scaleUpThreshold > 1.0) {
+            throw new IllegalArgumentException(
+                    "scaleUpThreshold must be between 0.0 and 1.0: " + scaleUpThreshold);
+        }
+        if (scaleDownThreshold >= scaleUpThreshold) {
+            throw new IllegalArgumentException(
+                    "scaleDownThreshold must be less than scaleUpThreshold: " +
+                    scaleDownThreshold + " >= " + scaleUpThreshold);
+        }
+        if (maxRampUpStep <= 0) {
+            throw new IllegalArgumentException("maxRampUpStep must be > 0: " + maxRampUpStep);
+        }
+        if (maxRampDownStep <= 0) {
+            throw new IllegalArgumentException("maxRampDownStep must be > 0: " + maxRampDownStep);
+        }
+        if (scalingPatienceCycles < 0) {
+            throw new IllegalArgumentException("scalingPatienceCycles must be >= 0: " + scalingPatienceCycles);
+        }
+
+        this.minChildren = minThreads;
+        this.maxChildren = maxThreads;
+        this.utilizationCheckPeriodNanos = windowUnit.toNanos(utilizationWindow);
+        this.scaleDownThreshold = scaleDownThreshold;
+        this.scaleUpThreshold = scaleUpThreshold;
+        this.maxRampUpStep = maxRampUpStep;
+        this.maxRampDownStep = maxRampDownStep;
+        this.scalingPatienceCycles = scalingPatienceCycles;
+    }
+
+    @Override
+    public EventExecutorChooser newChooser(EventExecutor[] executors) {
+        return new AutoScalingEventExecutorChooser(executors);
+    }
+
+    /**
+     * An immutable snapshot of the chooser's state. All state transitions
+     * are managed by atomically swapping this object.
+     */
+    private static final class AutoScalingState {
+        final int activeChildrenCount;
+        final long nextWakeUpIndex;
+        final EventExecutor[] activeExecutors;
+        final EventExecutorChooser activeExecutorsChooser;
+
+        AutoScalingState(int activeChildrenCount, long nextWakeUpIndex, EventExecutor[] activeExecutors) {
+            this.activeChildrenCount = activeChildrenCount;
+            this.nextWakeUpIndex = nextWakeUpIndex;
+            this.activeExecutors = activeExecutors;
+            this.activeExecutorsChooser = DefaultEventExecutorChooserFactory.INSTANCE.newChooser(activeExecutors);
+        }
+    }
+
+    private final class AutoScalingEventExecutorChooser implements EventExecutorChooser {
+        private final Runnable noOpTask = () -> { };
+        private final EventExecutor[] executors;
+        private final EventExecutorChooser allExecutorsChooser;
+        private final AtomicReference<AutoScalingState> state;
```

Comment:
use atomicReferenceUpdater?

Reply:
Honestly I don't think it really matters here as using the updater usually only makes sense to reduce memory overhead when you have a lot of these instances.